sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificInternalRow.scala

```diff
@@ -192,24 +192,41 @@ final class MutableAny extends MutableValue {
  */
 final class SpecificInternalRow(val values: Array[MutableValue]) extends BaseGenericInternalRow {
 
-  def this(dataTypes: Seq[DataType]) =
-    this(
-      dataTypes.map {
-        case BooleanType => new MutableBoolean
-        case ByteType => new MutableByte
-        case ShortType => new MutableShort
-        // We use INT for DATE internally
-        case IntegerType | DateType => new MutableInt
-        // We use Long for Timestamp internally
-        case LongType | TimestampType => new MutableLong
-        case FloatType => new MutableFloat
-        case DoubleType => new MutableDouble
-        case _ => new MutableAny
-      }.toArray)
+  private[this] def dataTypeToMutableValue(dataType: DataType): MutableValue = dataType match {
+    case BooleanType => new MutableBoolean
```

Comment:
Seems plausible. While micro-optimizing, I wonder if the ordering of these matches matters? is it faster to put more common types earlier?

Reply:
I think this is a good point.

Reply:
I will take a look.

Reply:
Did not see a lot of difference. Ran benchmarks on avro format with the following schema and 1024 * 1024 * 15 data points and 10 iterations.
```
root
|-- col1: struct (nullable = true)
|    |-- _1: integer (nullable = false)
|    |-- _2: integer (nullable = false)
```
With Integer type at bottom, getting between 7.1 to 7.7 seconds and with Integer type at top, getting between 7 and 7.5 seconds.

Reply:
OK. Might be worth reordering to integers, floating point, boolean, then the rest, but it's no big deal.

Reply:
Done

Reply:
Oh, I mean something more like:

Integer
Long
Float
Double
Boolean
Byte
Short

My guess is that's roughly in order of prevalence.